core/async/src/test_loop/event_handler.rs

```diff
@@ -0,0 +1,135 @@
+use std::time::Duration;
+
+use super::{delay_sender::DelaySender, multi_instance::IndexedLoopEventHandler};
+
+/// An event handler registered on a test loop. Each event handler usually
+/// handles only some events, so we will usually have multiple event handlers
+/// registered to cover all event types.
+pub trait LoopEventHandler<Data, Event> {
+    /// Called once, when the loop runs for the first time.
+    fn init(&mut self, _sender: DelaySender<Event>) {}
+
+    /// Handles an event. If this handler indeed handles the event, it should
+    /// return None after handling it. Otherwise, it should return Some with
+    /// the same event that was passed in, so that it can be given to the next
+    /// event handler.
+    fn handle(&mut self, event: Event, data: &mut Data) -> Option<Event>;
+}
+
+/// Functions that can be called on LoopEventHandler to adapt it to different
+/// situations.
+pub trait LoopEventHandlerHelpers<Data, Event> {
```

Comment:
do you expect this trait to be implemented for any other type than LoopEventHandler?

If not, how about just making it a method of LoopEventHandler? Or a standalone method?

Reply:
No, this is purely a helper for LoopEventHandler.

Making it a method of LoopEventHandler (which is a trait) results in the same syntax, however it is actually making a default implementation on a trait method. It allows the user to override the function, which makes no sense. That being said, it's not possible to override it because the return type has no public constructor. So I'll do it like you suggested, which saves a lot of boilerplate.